cpp/daal/src/externals/service_blas_ref.h

```diff
@@ -0,0 +1,255 @@
+    }
+
+    static void xxgemm(const char * transa, const char * transb, const DAAL_INT * p, const DAAL_INT * ny, const DAAL_INT * n, const double * alpha,
+                       const double * a, const DAAL_INT * lda, const double * y, const DAAL_INT * ldy, const double * beta, double * aty,
+                       const DAAL_INT * ldaty)
+    {
+        int old_threads = openblas_get_num_threads();
+        openblas_set_num_threads(1);
+        dgemm_(transa, transb, p, ny, n, alpha, a, lda, y, ldy, beta, aty, ldaty);
+        openblas_set_num_threads(old_threads);
+    }
+
+    static void xsymm(const char * side, const char * uplo, const DAAL_INT * m, const DAAL_INT * n, const double * alpha, const double * a,
+                      const DAAL_INT * lda, const double * b, const DAAL_INT * ldb, const double * beta, double * c, const DAAL_INT * ldc)
+    {
+        dsymm_(side, uplo, m, n, alpha, a, lda, b, ldb, beta, c, ldc);
+    }
+
+    static void xxsymm(char * side, char * uplo, DAAL_INT * m, DAAL_INT * n, double * alpha, double * a, DAAL_INT * lda, double * b, DAAL_INT * ldb,
+                       double * beta, double * c, DAAL_INT * ldc)
+    {
+        int old_threads = openblas_get_num_threads();
+        openblas_set_num_threads(1);
+        dsymm_(side, uplo, m, n, alpha, a, lda, b, ldb, beta, c, ldc);
+        openblas_set_num_threads(old_threads);
+    }
+
+    static void xgemv(const char * trans, const DAAL_INT * m, const DAAL_INT * n, const double * alpha, const double * a, const DAAL_INT * lda,
+                      const double * x, const DAAL_INT * incx, const double * beta, double * y, const DAAL_INT * incy)
+    {
+        dgemv_(trans, m, n, alpha, a, lda, x, incx, beta, y, incy);
+    }
+
+    static void xxgemv(const char * trans, const DAAL_INT * m, const DAAL_INT * n, const double * alpha, const double * a, const DAAL_INT * lda,
+                       const double * x, const DAAL_INT * incx, const double * beta, double * y, const DAAL_INT * incy)
+    {
+        int old_threads = openblas_get_num_threads();
+        openblas_set_num_threads(1);
+        dgemv_(trans, m, n, alpha, a, lda, x, incx, beta, y, incy);
+        openblas_set_num_threads(old_threads);
+    }
+
+    static void xaxpy(DAAL_INT * n, double * a, double * x, DAAL_INT * incx, double * y, DAAL_INT * incy) { daxpy_(n, a, x, incx, y, incy); }
+
+    static void xxaxpy(const DAAL_INT * n, const double * a, const double * x, const DAAL_INT * incx, double * y, const DAAL_INT * incy)
+    {
+        int old_threads = openblas_get_num_threads();
+        openblas_set_num_threads(1);
+        daxpy_(n, a, x, incx, y, incy);
+        openblas_set_num_threads(old_threads);
+    }
+
+    static double xxdot(const DAAL_INT * n, const double * x, const DAAL_INT * incx, const double * y, const DAAL_INT * incy)
+    {
+        int old_threads = openblas_get_num_threads();
+        openblas_set_num_threads(1);
```

Comment:
It looks like it will not work properly for xxgemm which should work simultaneously on all threads. Thus, locking at the begining and unlocking at the end will result in running xxgemm on a single thread.

Reply:
I agree with @KulikovNikita here. Using RAII approach would be better. Maybe the `_lock` in the name of the `omp_thread_lock` class is misleading.
The goal is to leave the execution logic the same as you have now in `xx` functions. I.e. set number of threads to 1 at the beginning, call OpenBLAS and restore the original number of threads after the call.
But let's hide this logic inside some helper RAII class (`omp_thread_lock` in Nikita's proposal). This is done in order not to accidently forget to restore the number of threads after the OpenBLAS call. Because it will happen automatically in the destructor of that RAII object when it goes out of the scope.

Reply:
Done